patches/api/0421-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,2335 @@
++    }
++
++    /**
++     * Creates a generic {@link DamageOriginBuilder} with optionally
++     * a position.
++     * Source position is used for explosion related damage.
++     *
++     * @param type           The damage type
++     * @param sourcePosition The source position
++     *
++     * @return A builder to then build a new {@link DamageOrigin}
++     */
++    @NotNull
++    static DamageOriginBuilder of(@NotNull DamageType type, @Nullable Position sourcePosition) {
++        return of(type, null, null, sourcePosition);
++    }
++
++    /**
++     * Creates a generic {@link DamageOriginBuilder}.
++     *
++     * @param type The damage type
++     *
++     * @return A builder to then build a new {@link DamageOrigin}
++     */
++    @NotNull
++    static DamageOriginBuilder of(@NotNull DamageType type) {
++        return of(type, null, null, null);
++    }
++
++    /**
++     * Gets the type of this damage origin.
++     * The type hold a lot of generic information like
++     * the hurt sound, death message format and more...
++     *
++     * @return the damage type
++     */
++    @NotNull
++    DamageType getType();
++
++    /**
++     * Checks if this damage origin is considered as sweep damage.
++     * A sweep damage generally hits multiple targets in a radius
++     * close to the attacker
++     *
++     * @return true for sweep damage
++     */
++    boolean isSweep();
++
++    /**
++     * Checks if this damage origin will do a critical hit or not.
++     * A critical hit will generally deal more damage to the target
++     *
++     * @return True for a critical damage
++     */
++    boolean isCritical();
++
++    /**
++     * Checks if this damage origin amount will be scaled
++     * to the difficulty of the world.
++     * In vanilla, this concern only the players.
++     *
++     * @return True for scalable damage
++     */
++    boolean willScalesWithDifficulty();
++
++    /**
++     * Checks if this damage origin is indirect
++     * Projectile launched by a player are qualified of
++     * indirect in this context for example. But it happens
++     * more generally when the direct entity ({@link #getDirectSource()})
++     * is different that the source entity ({@link #getSource()})
++     */
++    boolean isIndirect();
++
++    /**
++     * Checks if this damage origin is raw.
++     * The properties of a raw damage origin will
++     * only rely on its damage type (except for the
++     * entity, direct entity and source position).
++     *
++     * This includes all damage origin formed in
++     * vanilla ({@link #isVanilla()}) plus the plugin
++     * one without using the builder.
++     */
++    boolean isRaw();
++
++    /**
++     * Checks if this damage origin was constructed
++     * by vanilla and not by plugin. Note
++     * damage created using the damage command
++     * used by a plugin is considered as vanilla made!
++     */
++    boolean isVanilla();
```

Comment:
These methods are generally a bit confusing, in what context would you need to know if it was vanilla created? 
You should just be able to get the key from the damage type to determine if its custom damage.